### PR TITLE
[SPARK-12700] [SQL] embed condition into SMJ and BroadcastHashJoin

### PR DESCRIPTION
**Pull request by davies**
Currently SortMergeJoin and BroadcastHashJoin do not support condition, the need a followed Filter for that, the result projection to generate UnsafeRow could be very expensive if they generate lots of rows and could be filtered mostly by condition.

This PR brings the support of condition for SortMergeJoin and BroadcastHashJoin, just like other outer joins do.

This could improve the performance of Q72 by 7x (from 120s to 16.5s).

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala [hunk 1]
```diff
@@ -78,8 +78,11 @@ trait HashOuterJoin {
 
   @transient private[this] lazy val leftNullRow = new GenericInternalRow(left.output.length)
   @transient private[this] lazy val rightNullRow = new GenericInternalRow(right.output.length)
-  @transient private[this] lazy val boundCondition =
+  @transient private[this] lazy val boundCondition = if (condition.isDefined) {
```
Comment (nongli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala]: why did you change this?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala]: It's not related to this PR, just make it to aligned with others (a little faster for empty condition).

**Comment by davies** [comment]
@nongli Do you have more comments on this one?

**Comment by nongli** [comment]
LGTM. Let's make sure we don't need this when whole stage codegen is further along.

**Comment by davies** [comment]
For SMJ, we may not support whole stage codegen, this will be useful. For BroadcastHashJoin, once whole stage codegen is supported, we should not reply on this.

Merging this into master, thanks.

**Comment by nongli** [comment]
@davies I think it's reasonable to codegen the merge part of SMJ and have this start a new pipeline for whole stage codegen. Not required for the first go but shouldn't be too far away.

**Comment by davies** [comment]
@nongli That sounds good.
